File: shared/llm/transports/wire_trace.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any, TypedDict

from shared.env_config import env_flag, env_text
from shared.logging import logger

_TRACE_STRING_LIMIT = 8192
_IMAGE_DATA_PLACEHOLDER = "[image base64 omitted from wire trace]"
_SENSITIVE_WIRE_KEYS = frozenset(
    {
        "apikey",
        "xapikey",
        "token",
        "accesstoken",
        "refreshtoken",
        "secret",
        "password",
        "authorization",
        "cookie",
        "sessionid",
        "sessiontoken",
    }
)


class WireTraceRecord(TypedDict, total=False):
    ts: str
    kind: str
    session_id: str
    turn_id: str
    step: int
    attempt: int
    provider: str
    endpoint: str
    model: str
    timeout_s: int
    request_headers: dict[str, Any]
    request_payload: dict[str, Any]
    status_code: int
    response_headers: dict[str, Any]
    event: str
    data: str
    event_count: int
    ok: bool
    error: str


@dataclass(frozen=True, slots=True)
class WireTraceConfig:
    enabled: bool
    trace_dir: Path


@dataclass(slots=True)
class WireTraceContext:
    session_id: str
    turn_id: str
    step: int
    attempt: int
    provider: str
    turn_started_at: float = 0.0
    trace_path: str = ""
    turn_dir: str = ""


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[3]


def _now_iso() -> str:
    return datetime.now().astimezone().isoformat(timespec="milliseconds")


def _normalize_key(key: Any) -> str:
    return "".join(ch for ch in str(key or "").strip().lower() if ch.isalnum())


def _truncate_text(value: str, *, limit: int = _TRACE_STRING_LIMIT) -> str:
    text = str(value or "")
    if limit <= 0 or len(text) <= limit:
        return text

    marker = f"...[omitted {max(1, len(text) - limit)} chars]..."
    available = max(2, limit - len(marker))
    head_len = max(1, available // 2)
    tail_len = max(1, available - head_len)
    omitted = max(1, len(text) - head_len - tail_len)
    marker = f"...[omitted {omitted} chars]..."
    available = max(2, limit - len(marker))
    head_len = max(1, available // 2)
    tail_len = max(1, available - head_len)
    return f"{text[:head_len]}{marker}{text[-tail_len:]}"


def _is_image_block(value: Any) -> bool:
    return str(dict(value or {}).get("type") or "").strip().lower() == "image" and isinstance(
        dict(value or {}).get("source"), dict
    )


def _is_image_source(value: Any) -> bool:
    item = dict(value or {})
    media_type = str(item.get("media_type") or item.get("mimeType") or "").strip().lower()
    source_type = str(item.get("type") or "").strip().lower()
    return source_type == "base64" and "data" in item and media_type.startswith("image/")


def _sanitize_image_source(value: Any) -> dict[str, Any]:
    item = dict(value or {})
    return {
        "type": str(item.get("type") or "base64").strip() or "base64",
        "media_type": str(item.get("media_type") or item.get("mimeType") or "").strip(),
        "data": _IMAGE_DATA_PLACEHOLDER,
    }


def sanitize_wire_payload(
    value: Any,
    *,
    string_limit: int = _TRACE_STRING_LIMIT,
    _current_key: str = "",
    _seen: set[int] | None = None,
) -> Any:
    normalized_key = _normalize_key(_current_key)
    if normalized_key in _SENSITIVE_WIRE_KEYS:
        return "***"

    if value is None or isinstance(value, (bool, int, float)):
        return value

    if isinstance(value, str):
        return _truncate_text(value, limit=string_limit)

    if isinstance(value, (bytes, bytearray)):
        return _truncate_text(value.decode("utf-8", errors="replace"), limit=string_limit)

    seen = _seen if _seen is not None else set()
    if isinstance(value, dict):
        if _is_image_block(value):
            item = dict(value or {})
            return {
                "type": "image",
                "source": _sanitize_image_source(item.get("source")),
            }
        if _is_image_source(value):
            return _sanitize_image_source(value)
        obj_id = id(value)
        if obj_id in seen:
            return "[recursive]"
        seen.add(obj_id)
        try:
            return {
                str(key): sanitize_wire_payload(
                    item,
                    string_limit=string_limit,
                    _current_key=str(key),
                    _seen=seen,
                )
                for key, item in value.items()
            }
        finally:
            seen.discard(obj_id)

    if isinstance(value, (list, tuple, set)):
        obj_id = id(value)
        if obj_id in seen:
            return "[recursive]"
        seen.add(obj_id)
        try:
            return [
                sanitize_wire_payload(
                    item,
                    string_limit=string_limit,
                    _seen=seen,
                )
                for item in value
            ]
        finally:
            seen.discard(obj_id)

    return _truncate_text(str(value), limit=string_limit)


def load_wire_trace_config() -> WireTraceConfig:
    trace_dir_raw = env_text("AGENT_SSE_WIRE_TRACE_DIR", "logs/sse_wire_traces")
    trace_dir = Path(trace_dir_raw or "logs/sse_wire_traces")
    if not trace_dir.is_absolute():
        trace_dir = (_repo_root() / trace_dir).resolve()
    return WireTraceConfig(
        enabled=env_flag("AGENT_SSE_WIRE_TRACE_ENABLED", True),
        trace_dir=trace_dir,
    )


def dated_session_trace_dir(
    base_dir: Path,
    *,
    session_id: str,
    now_ts: float | None = None,
) -> Path:
    safe_session = str(session_id or "").strip()
    moment = time.time() if now_ts is None else float(now_ts)
    dated_dir = base_dir / time.strftime("%Y%m%d", time.localtime(moment))
    if not safe_session:
        return dated_dir

    hour_minute = time.strftime("%H%M", time.localtime(moment))
    return dated_dir / f"{hour_minute}_{safe_session}"


def wire_trace_turn_dir(
    wire_config: WireTraceConfig,
    *,
    session_id: str,
    turn_id: str,
    started_at: float | None = None,
) -> str:
    if not wire_config.enabled:
        return ""
    session = str(session_id or "").strip()
    turn = str(turn_id or "").strip()
    if not session or not turn:
        return ""
    session_dir = dated_session_trace_dir(
        wire_config.trace_dir,
        session_id=session,
        now_ts=started_at,
    )
    return str(session_dir / turn)


class WireTraceWriter:
    def __init__(
        self,
        *,
        context: WireTraceContext,
        wire_config: WireTraceConfig | None = None,
    ) -> None:
        self.context = context
        self.config = wire_config or load_wire_trace_config()
        self._enabled = bool(self.config.enabled)
        self._handle = None
        self._event_count = 0

        if not self._enabled:
            return

        try:
            turn_dir_raw = wire_trace_turn_dir(
                self.config,
                session_id=self.context.session_id,
                turn_id=self.context.turn_id,
                started_at=self.context.turn_started_at,
            )
            if not turn_dir_raw:
                self._enabled = False
                return
            turn_dir = Path(turn_dir_raw)
            turn_dir.mkdir(parents=True, exist_ok=True)
            path = turn_dir / f"step_{self.context.step}_attempt_{self.context.attempt}.jsonl"
            self._handle = path.open("a", encoding="utf-8")
            self.context.turn_dir = str(turn_dir)
            self.context.trace_path = str(path)
        except Exception as error:
            logger.warning("[WireTrace] failed to initialize wire trace file: %s", error)
            self._enabled = False

    @property
    def trace_path(self) -> str:
        return self.context.trace_path

    @property
    def turn_dir(self) -> str:
        return self.context.turn_dir

    @property
    def event_count(self) -> int:
        return self._event_count

    def write_request_start(
        self,
        *,
        endpoint: str,
        model: str,
        timeout_s: int,
        request_headers: dict[str, Any],
        request_payload: dict[str, Any],
    ) -> None:
        self._write(
            {
                "ts": _now_iso(),
                "kind": "request_start",
                "session_id": self.context.session_id,
                "turn_id": self.context.turn_id,
                "step": self.context.step,
                "attempt": self.context.attempt,
                "provider": self.context.provider,
                "endpoint": str(endpoint or "").strip(),
                "model": str(model or "").strip(),
                "timeout_s": int(timeout_s),
                "request_headers": sanitize_wire_payload(dict(request_headers or {})),
                "request_payload": sanitize_wire_payload(dict(request_payload or {})),
            }
        )

    def write_response_start(
        self,
        *,
        status_code: int,
        response_headers: dict[str, Any],
    ) -> None:
        self._write(
            {
                "ts": _now_iso(),
                "kind": "response_start",
                "session_id": self.context.session_id,
                "turn_id": self.context.turn_id,
                "step": self.context.step,
                "attempt": self.context.attempt,
                "provider": self.context.provider,
                "status_code": int(status_code),
                "response_headers": sanitize_wire_payload(dict(response_headers or {})),
            }
        )

    def write_wire_event(self, *, event_name: str, raw_data: str) -> None:
        self._event_count += 1
        self._write(
            {
                "ts": _now_iso(),
                "kind": "wire_event",
                "session_id": self.context.session_id,
                "turn_id": self.context.turn_id,
                "step": self.context.step,
                "attempt": self.context.attempt,
                "provider": self.context.provider,
                "event": str(event_name or "").strip() or "message",
                "data": str(raw_data or ""),
            }
        )

    def write_parse_error(self, *, event_name: str, raw_data: str, error: Exception) -> None:
        self._write(
            {
                "ts": _now_iso(),
                "kind": "parse_error",
                "session_id": self.context.session_id,
                "turn_id": self.context.turn_id,
                "step": self.context.step,
                "attempt": self.context.attempt,
                "provider": self.context.provider,
                "event": str(event_name or "").strip() or "message",
                "data": str(raw_data or ""),
                "error": str(error or "").strip() or type(error).__name__,
            }
        )

    def write_request_end(self, *, ok: bool, error: str = "") -> None:
        self._write(
            {
                "ts": _now_iso(),
                "kind": "request_end",
                "session_id": self.context.session_id,
                "turn_id": self.context.turn_id,
                "step": self.context.step,
                "attempt": self.context.attempt,
                "provider": self.context.provider,
                "ok": bool(ok),
                "event_count": self._event_count,
                "error": str(error or "").strip(),
            }
        )

    def close(self) -> None:
        if self._handle is None:
            return
        try:
            self._handle.close()
        finally:
            self._handle = None

    def _write(self, record: WireTraceRecord) -> None:
        if not self._enabled or self._handle is None:
            return
        try:
            self._handle.write(json.dumps(dict(record or {}), ensure_ascii=False) + "\n")
            self._handle.flush()
        except Exception as error:
            logger.warning("[WireTrace] failed to write wire trace record: %s", error)


__all__ = [
    "WireTraceConfig",
    "WireTraceContext",
    "WireTraceRecord",
    "WireTraceWriter",
    "load_wire_trace_config",
    "sanitize_wire_payload",
    "wire_trace_turn_dir",
]

File: packages/agent/runtime/src/providers/wire-trace.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import type { JsonObject, JsonValue } from "@lxe/protocol";
import {
  createLogger,
  envFlag,
  envText,
  sanitizeLogValueWithPolicy,
  type Environment,
} from "@lxe/core";

const ABSOLUTE_PATH = /(?:[A-Za-z]:\\|\/(?:home|Users|var|tmp)\/)[^\s"']+/gu;
const WIRE_STRING_LIMIT = 8_192;
const WIRE_COLLECTION_LIMIT = 10_000;
const logger = createLogger("runtime.wire_trace");
const localDay = (date: Date): string =>
  `${date.getFullYear()}${String(date.getMonth() + 1).padStart(2, "0")}${String(date.getDate()).padStart(2, "0")}`;
const localIso = (date = new Date()): string => {
  const offsetMinutes = -date.getTimezoneOffset();
  const sign = offsetMinutes >= 0 ? "+" : "-";
  const offset = Math.abs(offsetMinutes);
  return [
    `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-${String(date.getDate()).padStart(2, "0")}`,
    `T${String(date.getHours()).padStart(2, "0")}:${String(date.getMinutes()).padStart(2, "0")}:${String(date.getSeconds()).padStart(2, "0")}.${String(date.getMilliseconds()).padStart(3, "0")}`,
    `${sign}${String(Math.floor(offset / 60)).padStart(2, "0")}:${String(offset % 60).padStart(2, "0")}`,
  ].join("");
};

const truncateWireText = (value: string): string => {
  if (value.length <= WIRE_STRING_LIMIT) return value;
  let marker = `...[omitted ${Math.max(1, value.length - WIRE_STRING_LIMIT)} chars]...`;
  let available = Math.max(2, WIRE_STRING_LIMIT - marker.length);
  let headLength = Math.max(1, Math.floor(available / 2));
  let tailLength = Math.max(1, available - headLength);
  const omitted = Math.max(1, value.length - headLength - tailLength);
  marker = `...[omitted ${omitted} chars]...`;
  available = Math.max(2, WIRE_STRING_LIMIT - marker.length);
  headLength = Math.max(1, Math.floor(available / 2));
  tailLength = Math.max(1, available - headLength);
  return `${value.slice(0, headLength)}${marker}${value.slice(-tailLength)}`;
};

const normalizedKey = (value: string): string => value.toLowerCase().replaceAll(/[^a-z0-9]+/gu, "");
const textLength = (value: unknown): number => typeof value === "string" ? value.length : 0;

const prepareWireValue = (
  value: unknown,
  key = "",
  parentType = "",
  seen = new WeakSet<object>(),
): unknown => {
  const normalized = normalizedKey(key);
  if (normalized === "signature" || normalized === "signaturedelta") return "***";
  if (parentType === "redacted_thinking" && normalized === "data") {
    return `[redacted thinking data omitted: ${textLength(value)} chars]`;
  }
  if (value === null || value === undefined || typeof value === "boolean" || typeof value === "number") return value;
  if (typeof value === "string") return truncateWireText(value.replaceAll(ABSOLUTE_PATH, "<path>"));
  if (typeof value !== "object") return value;
  if (value instanceof Error || value instanceof Date || value instanceof URL) return value;
  if (value instanceof ArrayBuffer || ArrayBuffer.isView(value)) return value;
  if (value instanceof Map || value instanceof Set) return value;
  if (seen.has(value)) return "[recursive]";
  seen.add(value);
  try {
    if (Array.isArray(value)) return value.map((item) => prepareWireValue(item, "", parentType, seen));
    const source = value as Record<string, unknown>;
    const type = String(source.type ?? parentType).trim().toLowerCase();
    const imageSource = type === "base64"
      && String(source.media_type ?? source.mimeType ?? "").trim().toLowerCase().startsWith("image/");
    return Object.fromEntries(Object.entries(source).map(([childKey, item]) => {
      if (imageSource && normalizedKey(childKey) === "data") {
        return [childKey, `[image base64 omitted: ${textLength(item)} chars]`];
      }
      return [childKey, prepareWireValue(item, childKey, type, seen)];
    }));
  } catch (error) {
    return `[unserializable: ${error instanceof Error ? error.message : String(error)}]`;
  } finally {
    seen.delete(value);
  }
};

export function sanitizeWireTraceValue(value: unknown): JsonValue {
  return sanitizeLogValueWithPolicy(prepareWireValue(value), {
    maxDepth: 32,
    maxItems: WIRE_COLLECTION_LIMIT,
    maxString: WIRE_STRING_LIMIT,
    maxStackString: WIRE_STRING_LIMIT,
  }) as JsonValue;
}

const sanitizeRawWireData = (value: string): string => {
  try {
    return JSON.stringify(sanitizeWireTraceValue(JSON.parse(value)));
  } catch {
    return truncateWireText(value
      .replace(/\bBearer\s+[A-Za-z0-9._~+\/-]+=*/gi, "Bearer [redacted]")
      .replace(
        /((?:"?)(?:authorization|api[_-]?key|access[_-]?token|refresh[_-]?token|token|secret|cookie|password|signature)(?:"?)\s*[:=]\s*)(?:"[^"]*"|'[^']*'|[^\s,;]+)/gi,
        "$1\"***\"",
      ));
  }
};

export interface RuntimeWireTraceAttemptContext {
  step: number;
  attempt: number;
  provider: string;
  model: string;
  endpoint: string;
  timeoutMs: number;
}

export interface RuntimeWireTraceAttempt {
  requestStart(requestHeaders: JsonObject, requestPayload: JsonObject): void;
  responseStart(statusCode: number, responseHeaders: JsonObject): void;
  event(eventName: string, payload: unknown): void;
  parseError(eventName: string, rawData: string, error: unknown): void;
  end(ok: boolean, error?: string): void;
}

export interface RuntimeWireTraceTurn {
  startProviderAttempt(context: RuntimeWireTraceAttemptContext): RuntimeWireTraceAttempt | undefined;
}

export interface RuntimeWireTraceControllerPort {
  startTurn(sessionId: string, turnId: string): RuntimeWireTraceTurn;
}

interface WireTraceBaseRecord extends JsonObject {
  ts: string;
  kind: string;
  session_id: string;
  turn_id: string;
  step: number;
  attempt: number;
  provider: string;
}

class FileWireTraceAttempt implements RuntimeWireTraceAttempt {
  private failed = false;
  private requestStarted = false;
  private responseStarted = false;
  private ended = false;
  private eventCount = 0;

  constructor(
    private readonly path: string,
    private readonly sessionId: string,
    private readonly turnId: string,
    private readonly context: RuntimeWireTraceAttemptContext,
  ) {
    try {
      mkdirSync(dirname(path), { recursive: true });
    } catch (error) {
      this.failed = true;
      logger.warn("wire_trace_initialize_failed", {
        trace_path: path,
        step: context.step,
        attempt: context.attempt,
        provider: context.provider,
        error,
      });
    }
  }

  requestStart(requestHeaders: JsonObject, requestPayload: JsonObject): void {
    if (this.ended || this.requestStarted) return;
    this.requestStarted = true;
    this.write({
      ...this.base("request_start"),
      endpoint: sanitizeWireTraceValue(this.context.endpoint),
      model: sanitizeWireTraceValue(this.context.model),
      timeout_s: Math.max(0, Math.round(this.context.timeoutMs / 1_000)),
      request_headers: sanitizeWireTraceValue(requestHeaders),
      request_payload: sanitizeWireTraceValue(requestPayload),
    });
  }

  responseStart(statusCode: number, responseHeaders: JsonObject): void {
    if (this.ended || this.responseStarted) return;
    this.responseStarted = true;
    this.write({
      ...this.base("response_start"),
      status_code: Math.max(0, Math.trunc(statusCode)),
      response_headers: sanitizeWireTraceValue(responseHeaders),
    });
  }

  event(eventName: string, payload: unknown): void {
    if (this.ended) return;
    try {
      const data = JSON.stringify(sanitizeWireTraceValue(payload)) ?? "null";
      this.eventCount += 1;
      this.write({
        ...this.base("wire_event"),
        event: sanitizeWireTraceValue(eventName.trim() || "message"),
        data,
      });
    } catch (error) {
      this.parseError(eventName, "", error);
    }
  }

  parseError(eventName: string, rawData: string, error: unknown): void {
    if (this.ended) return;
    this.write({
      ...this.base("parse_error"),
      event: sanitizeWireTraceValue(eventName.trim() || "message"),
      data: sanitizeRawWireData(rawData),
      error: sanitizeWireTraceValue(error instanceof Error ? error.message : String(error)),
    });
  }

  end(ok: boolean, error = ""): void {
    if (this.ended) return;
    this.ended = true;
    this.write({
      ...this.base("request_end"),
      ok,
      event_count: this.eventCount,
      error: sanitizeWireTraceValue(error.trim()),
    });
  }

  private base(kind: string): WireTraceBaseRecord {
    return {
      ts: localIso(),
      kind,
      session_id: this.sessionId,
      turn_id: this.turnId,
      step: this.context.step,
      attempt: this.context.attempt,
      provider: this.context.provider,
    };
  }

  private write(record: JsonObject): void {
    if (this.failed) return;
    try {
      appendFileSync(this.path, `${JSON.stringify(record)}\n`, "utf8");
    } catch (error) {
      this.failed = true;
      logger.warn("wire_trace_write_failed", {
        trace_path: this.path,
        trace_kind: record.kind,
        step: this.context.step,
        attempt: this.context.attempt,
        provider: this.context.provider,
        error,
      });
    }
  }
}

class FileRuntimeWireTraceTurn implements RuntimeWireTraceTurn {
  constructor(
    private readonly sessionId: string,
    private readonly turnId: string,
    private readonly wireDirectory?: string,
  ) {}

  startProviderAttempt(context: RuntimeWireTraceAttemptContext): RuntimeWireTraceAttempt | undefined {
    if (!this.wireDirectory) return undefined;
    const safeStep = Math.max(0, Math.trunc(context.step));
    const safeAttempt = Math.max(1, Math.trunc(context.attempt));
    return new FileWireTraceAttempt(
      join(this.wireDirectory, `step_${safeStep}_attempt_${safeAttempt}.jsonl`),
      this.sessionId,
      this.turnId,
      { ...context, step: safeStep, attempt: safeAttempt },
    );
  }
}

export function configureRuntimeWireTracing(options: {
  projectRoot: string;
  stateRoot?: string;
  environment: Environment;
}): RuntimeWireTraceControllerPort {
  const enabled = envFlag(options.environment, "LOCAL_LOGS_ENABLED", false);
  const wireEnabled = enabled && envFlag(options.environment, "AGENT_SSE_WIRE_TRACE_ENABLED", true);
  const explicitStateRoot = String(options.stateRoot ?? "").trim();
  const configuredRoot = explicitStateRoot ? resolve(explicitStateRoot) : resolve(options.projectRoot);
  const managedPrefix = explicitStateRoot ? "logs" : "var/logs";
  const wireRoot = resolve(
    configuredRoot,
    envText(options.environment, "AGENT_SSE_WIRE_TRACE_DIR", `${managedPrefix}/sse_wire_traces`),
  );
  return {
    startTurn: (sessionId, turnId) => {
      const now = new Date();
      const day = localDay(now);
      const time = now.toTimeString().slice(0, 5).replace(":", "");
      const safeSession = sessionId.replaceAll(/[^A-Za-z0-9_-]+/gu, "_") || "session";
      const safeTurn = turnId.replaceAll(/[^A-Za-z0-9_-]+/gu, "_") || "turn";
      const directoryName = `${time}_${safeSession}`;
      const wireDirectory = join(wireRoot, day, directoryName, safeTurn);
      return new FileRuntimeWireTraceTurn(
        sessionId,
        turnId,
        wireEnabled ? wireDirectory : undefined,
      );
    },
  };
}
